### ISSUE TITLE
Make hp_max and breath _max persistent (maybe all other properties too?)

### ISSUE THREAD

**Issue by Wuzzy2** [minetest/minetest#8211]
##### Issue type
 - Bug report
 
##### Minetest version
```
f5bdc04ab5b47bd094ba0ced5f086e8fe6b12298
```

##### Summary
When a player has any non-default max. breath or health, these values reset to the defaults after restarting the server. :-(

##### Steps to reproduce

Step 1: Use this simple mod so you can see your current hp/breath and max. values as raw numbers in your interface:

```
local test = {}
test.playerlist = {}

minetest.register_on_respawnplayer(function(player)
	local playername = player:get_player_name()
	test.hud_update(playername)
end)

minetest.register_on_joinplayer(function(player)
	local playername = player:get_player_name()
	
	if test.playerlist[playername] == nil then
		test.playerlist[playername] = {}
	end

	test.hud_add(playername)
end)


minetest.register_globalstep(function(dtime)
	for k,v in pairs(test.playerlist) do
		test.hud_update(k)
	end
end)

function test.get_string(player)
	local prop = player:get_properties()
	return string.format("Health: %s/%s\nBreath: %s/%s", player:get_hp(), prop.hp_max, player:get_breath(), prop.breath_max)
end

function test.hud_add(playername)
	local player = minetest.get_player_by_name(playername)
	local id = player:hud_add({
		hud_elem_type = "text",
		position = { x = 0, y=1 },
		text = test.get_string(player),
		scale = { x = 0, y = 0 },
		alignment = { x = 1, y = 0},
		direction = 1,
		number = 0xFFFFFF,
		offset = { x = 20, y = -103}
	})
	test.playerlist[playername].hudid = id
	return id
end

function test.hud_update(playername)
	local player = minetest.get_player_by_name(playername)
	player:hud_change(test.playerlist[playername].hudid, "text", test.get_string(player))
end

function test.hud_remove(playername)
	local player = minetest.get_player_by_name(playername)
	player:hud_remove(test.playerlist[playername].hudid)
end
```

Step 2: Use `luacmd` to set you max health or breath to any non-default value
Step 3: Restart server

Tested in singleplayer only.

**Comment by jastevenson303** [comment]
all properties are reset

**Comment by SmallJoker** [comment]
None of the player properties are saved. Not a bug IMO.

**Comment by Wuzzy2** [comment]
In that case, turn this issue into a documentation problem because the documentation does not seem to say that object properties are NOT persisted. This is not obvious and a major gotcha, so it definitely must be mentioned.

By the way: I would rather recommend using player attributes for saving such stuff as it's easier to do.

**Comment by Wuzzy2** [comment]
After thinking about it more, no this is not just a documentation, it's still bad design.

It's weird, strange and annoying that hp and breath are automatically persisted, but max hp and breath aren't. It's a weird inconsistency and ugly API design.

Apart from the ugliness (which is not really a bug), there is also a major problem with that:
Redundancy.

Suppose mod A messes with max. HP (let's say they increase it by 5) and persists it with player attribute.
Now mod B wants to mess with max. HP as well (let's say they increase it by 6 more HP points). But this time, they persist it with mod storage like SmallJoker suggested. There is no standard (!) way to persist max HP/breath after all.
Now we have the max HP persisted twice with 2 different values. In other words: Redundant! This sucks! When the server restarts, both mods try to restore the max HP according to “their” value, but only one of them will “win”. Classic race condition and clearly NOT what we wanted.
So unless someone can propose a standard (!) way to persist max HP/breath which doesn't break when you use more than 1 mod at once, the original issue remains.

Therefore, I think max HP and breath should be persisted automatically, just like HP and breath already are, with no further intervention from mods required. Anything else is just messy, confusing and leads to buggy code.

**Comment by paramat** [comment]
Max hp and max breath are player object properties, which are not persistent, making certain properties persist while others don't would be inconsistent and likely to cause problems and complexity.

Properties are always set by a mod on player join, so some other kind of persistence is not needed. The problem here is: "Use `luacmd` to set you max health or breath to any non-default value" which is a very rare way to set properties.

Not a bug but maybe documentation.

**Comment by Wuzzy2** [comment]
> Max hp and max breath are player object properties, which are not persistent, making certain properties persist while others don't would be inconsistent and likely to cause problems and complexity.

I agree. But then this begs the question why max HP/breath are player object properties in the first place. Breath and HP are not properties, but MAX breath and HP are? That doesn't sound very consistent to me.

> problem here is: "Use luacmd to set you max health or breath to any non-default value" which is a very rare way to set properties.

Look, of course this is not my actual *use case*! This is just part of my steps to reproduce to provide evidence in a very fast and simple reproducible way that the problem exists. Of course you could use any other method, like writing a test mod or whatever, but the end result is the same: `hp_max` and `breath_max` are not persisted.

The real use case would of course be something like: I want to write a game that sets max. HP of all players to 100 and it stays that way forever. I thought this “use case” is so obvious that I did not even mention it.

To summarize:

I don't consider this issue resolved. Nobody has yet shown a possible way or workaround to persist max. HP or max. breath reliably and without redundancy / data inconsistency. In my previous post I have shown that the proposed way of just using mod storage does not work. This point has been ignored so far.

Therefore, I consider it currently *impossible* to persist max. HP and breath. If max. HP and breath are not persisted between server restarts, nor is there a known workaround for it, the feature is currently **broken by design**.

**Comment by jastevenson303** [comment]
Did you try player attributes, then, if mod storage isn't for you?

**Comment by Linuxdirk** [comment]
> the feature is currently **broken by design**.

So better store stuff in player meta instead if player properties?

**Comment by jastevenson303** [comment]
properties aren't meant to persist, it would seem -- whereas meta is?
interesting.

**Comment by Linuxdirk** [comment]
> properties aren't meant to persist, it would seem -- whereas meta is?

Yes, I can conform that meta is persistent. Properties should be persistent, too. Or it should be stated clearly that properties are not persistent. I'd love to store stuff in the properties, it would make much more sense for certain things. But persistency is a must-have for that.

Using mod storage or meta data for persistent storage of player properties (and loading and setting them in an on_joinplayer hook) are only a workaround.

**Comment by paramat** [comment]
> Breath and HP are not properties, but MAX breath and HP are?

Makes complete sense. There are many properties related to non-property values.
> Look, of course this is not my actual _use case_! This is just part of my steps to reproduce to provide evidence in a very fast and simple reproducible way that the problem exists.

Yes :) But any mod that sets properties already acts 'on joinplayer' setting values that can be stored in mod storage, so there's already the ability to persist. You were using a highly unusual and rare way of setting max HP/breath to claim there is a bug, that's what i am pointing out.
> The real use case would of course be something like: I want to write a game that sets max. HP of all players to 100 and it stays that way forever.

You can do this already.
Making properties persist is extra complexity, will probably cause problems and so not be worth it since it's already possible.
> Using mod storage or meta data for persistent storage of player properties (and loading and setting them in an on_joinplayer hook) are only a workaround.

Not a workaround, that's the expected method to use.

We could consider making all properties persist in case it turns out to be a good idea, but low priority.

**Comment by SmallJoker** [comment]
#8914 documented that object properties are not persistent (includes max hp/breath). Hence I assume this is a feature request to make all properties persistent.
Labels updated.

**Comment by Wuzzy2** [comment]
Well, I see that meta can be used by mods to somehow persist the max breath and hp anyway. So I no longer think this is a pressing issue, nor something broken by design.
